File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    printf("Number: ");
    long long cc_number = get_long_long();
    long long cc_temp = cc_number;
    long long sum = 0;
    
    while (cc_temp > 0)
    {
        int odd = (cc_temp % 100) / 10;
        odd = odd * 2;
        if (odd > 9) {
            int rem = odd % 10;
            odd = rem + 1;
        }
        sum += odd;
        
        int even = cc_temp % 10;
        sum += even;
        
        cc_temp /= 100;
    }

    cc_temp = cc_number;
    int length = 0;
    int first_two;
    while (cc_temp > 0)
    {
        if (cc_temp < 100 && cc_temp > 9)
        {
            first_two = cc_temp;
        }
        cc_temp /= 10;
        length++;
    }
    
    if (sum % 10 == 0)
    {
        switch (first_two)
        {
            case 34:
            case 37:
                if (length == 15)
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
            case 51:
            case 52:
            case 53:
            case 54:
            case 55:
                if (length == 16)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
            default:
                if (first_two / 10 == 4 && (length == 13 || length == 16))
                {
                    printf("VISA\n");
                }
                else
                {
                    printf("INVALID\n");
                }
                break;
        }
    }
    else
    {
        printf("INVALID\n");
    }
    
    return 0;
}

File: pset6/credit.py
import cs50

def main():
    print("Number: ", end = "")
    
    # convert to string
    string = str(cs50.get_int())
    num = len(string)
    
    # start from second last digit
    index = num - 2
    total = 0
    
    # loop through every other digit from the end
    while (index >= 0):
        # get digit at index position of the string
        temp = 2 * int(string[index])
        
        # add the sum of its digits to total
        if (temp < 10):
            total += temp
        else:
            total += 1 + (temp % 10)
        index -= 2
    
    # loop through all the other digits and add to running total
    index = num - 1
    while (index >=0):
        total += int(string[index])
        index -= 2
    
    # checksum
    if (total % 10 == 0):
        d = int(string[:2])
        print(d)
        if ((d == 34 or d == 37) and num == 15):
            print("AMEX")
        elif (51 <= d and d <= 55 and num == 16):
            print("MASTERCARD")
        elif (int(d / 10) == 4 and (num == 13 or num == 16)):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")

if __name__ == "__main__":
    main()